### PR TITLE
Create custom resource for database user initialization on RDS

### PR DESCRIPTION
**Pull request by tolgatakir**

### CODE REVIEW

File: lambdas/src/handlers/database-user-initializer/update.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                script = script.concat(await ScriptBuilder.dropUserScript(removedUser));
+            }
+
+            const addedUsers: any = _.differenceBy(props.DatabaseUsers, oldProps.DatabaseUsers, 'username');
+            for (const addedUser of addedUsers) {
+                script = script.concat(await ScriptBuilder.createUserScript(addedUser, credential.dbname, props.Region));
+            }
+
+            for (const newUser of props.DatabaseUsers) {
+                for (const oldUser of oldProps.DatabaseUsers) {
```
Comment (rwxdash) [inline lambdas/src/handlers/database-user-initializer/update.ts]: Is it possible to do this without having to do a nested loop? Could it be possible to handle this by using `indexOf` over the object's `username` field?

File: src/database-user-initializer.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     *
+     */
+    readonly databaseUsers: (User | IAMUser)[];
+}
+
+/**
+ *
+ */
+export class DatabaseUserInitializer extends cdk.Resource {
```
Comment (serkan-ozal) [inline src/database-user-initializer.ts:57]: A `README` file is required to explain how to use this `DatabaseUserInitializer` custom resource in AWS CDK app

File: src/database-script-runner.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                DatabaseAdminUserSecretName: this.databaseAdminUserSecret.secretName,
+                Script: this.script
+            },
+        });
+    }
+
+    protected validate(): string[] {
+        const errors: string[] = [];
+
+        if (this.prefix.trim().length === 0) {
```
Comment (serkan-ozal) [inline src/database-script-runner.ts]: Should we check whether `prefix` is not `null`/`undefined` here?

File: src/database-script-runner.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            errors.push('Prefix properties must not be empty.');
+        }
+
+        if (this.script.trim().length === 0) {
```
Comment (serkan-ozal) [inline src/database-script-runner.ts]: Should we check whether `script` is not `null`/`undefined` here?

File: src/database-user-initializer.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+                Region: cdk.Stack.of(this).region,
+                DatabaseAdminUserSecretName: this.databaseAdminUserSecret.secretName,
+                DatabaseUsers: paramater
+            },
+        });
+    }
+
+    protected validate(): string[] {
+        const errors: string[] = [];
+        if (this.prefix.trim().length === 0) {
```
Comment (serkan-ozal) [inline src/database-user-initializer.ts]: Should we check whether `prefix` is not `null`/`undefined` here?

File: src/database-user-initializer.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            errors.push('Prefix properties must not be empty.');
+        }
+
+        if (this.databaseUsers.length === 0) {
```
Comment (serkan-ozal) [inline src/database-user-initializer.ts]: Should we check whether `databaseUsers` is not `null`/`undefined` here?

File: src/database-script-runner.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     *
+     */
+    readonly script: string;
+}
+
+/**
+ *
+ */
+export class DatabaseScriptRunner extends cdk.Resource {
```
Comment (serkan-ozal) [inline src/database-script-runner.ts:32]: A `README` file is required to explain how to use this `DatabaseScriptRunner` custom resource in AWS CDK app

File: package.json [hunk 8]
```diff
@@ -0,0 +1,57 @@
+{
+  "name": "@thundra/cdk-rds-initializer",
+  "version": "0.0.1",
+  "description": "Lambda handler for running custom database script",
```
Comment (serkan-ozal) [inline package.json]:
```
AWS CDK custom resource to initialize AWS RDS database
```

File: lambdas/src/handlers/script-builder.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
+import * as _ from "lodash";
+import {SecretsManager} from "./secrets-manager";
+import {User} from "./user";
+
+export class ScriptBuilder {
```
Comment (serkan-ozal) [inline lambdas/src/handlers/script-builder.ts:5]: Are all these SQL scripts are specific to MySQL or can they also be run by PostgreSQL?

**Review by rwxdash — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -1 +1,280 @@
-# thundra-cdk-rds-initializer
\ No newline at end of file
+# @thundra/cdk-rds-initializer
+## RDS Database User Initializer and Script Runner Construct for AWS CDK
+
+This package provides Constructs for creating database user & running database script which can be used in RDS.
+
+This package automatically create database user and run database script.
```
Comment (serkan-ozal) [inline README.md:6]:
```
This package automatically creates specified database user(s) and run given database script(s).
```

**Review by serkan-ozal — APPROVED** [review]

**Pull request merged**